src/Perlang.Interpreter/PerlangInterpreter.cs

```diff
@@ -948,10 +949,23 @@ private static bool IsEqual(object? a, object? b)
 
             if (a == null)
             {
+                // null != non-null
                 return false;
             }
 
-            return a.Equals(b);
+            // Note: must use dynamic here since `object == object` behaves differently than e.g. `int == double`,
+            // tragically enough. More details:
+            // https://learn.microsoft.com/en-us/dotnet/api/System.Object.Equals?view=net-7.0#notes-for-callers
```

Comment:
Here's a clickable version of that link: https://learn.microsoft.com/en-us/dotnet/api/System.Object.Equals?view=net-7.0#notes-for-callers

What's particularly unintuitive in C# is perhaps the difference between these:

```c#
int i = 12;
int j = 12;

object i2 = i;
object j2 = j;

bool b = i == j;   // true
bool c = i2 == j2; // false
```

I believe this is because of [boxing](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/types/boxing-and-unboxing), which is necessary to be able to store a value type in an `object` variable, but it's still hard to argue that the above provides an _extremely_ confusing developer experience. The interesting question is: would it have been reasonably possible to provide any alternative semantics that would have worked _better_ in this regard, without any other adverse effects? :thinking:

The thing is that C# is fairly advanced in this case; much more similar to C++ than e.g. Java, with support for overloading the `==` operator and things like this. That is a very nice thing for certain use cases, but I guess it also comes at a price.